pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,225 @@
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import torch_ones_like, torch_zeros_like
+from pyro.infer import TracePosterior
+from pyro.util import ng_ones, ng_zeros
+from tests.common import assert_equal
+
+
+def verlet_integrator(z, r, grad_potential, step_size, num_steps):
+    """
+    Velocity Verlet integrator.
+
+    :param z: dictionary of sample site names and their current values
+    :param r: dictionary of sample site names and corresponding momenta
+    :param grad_potential: function that returns gradient of the potential given z
+        for each sample site
+    :return: (z_next, r_next) having same types as (z, r)
+    """
+    # deep copy the current state - (z, r)
+    z_next = {key: val.clone().detach() for key, val in z.items()}
+    r_next = {key: val.clone().detach() for key, val in r.items()}
+    retain_grads(z_next)
+    retain_grads(r_next)
+    grads = grad_potential(z_next)
+
+    for _ in range(num_steps):
+        # detach graph nodes for next iteration
```

Comment:
I believe you should call `detach_grads()` *after* the update step. Otherwise gradients will be propagated backwards one step in time and never accessed, right?

Reply:
The gradient has been computed right before the start of the loop, cloned and stored in `grads`. This `detach` call is to ensure that we do not propagate that graph further.